src/weighted_map.h

```diff
@@ -0,0 +1,194 @@
+// removed from the map with probability in proportion to their weight. This is
+// used to implement mempool limiting specified in ZIP 401.
+//
+// In order to efficiently implement random selection by weight, we keep track
+// of the total weight of all keys in the map. For performance reasons, the
+// map is represented as a binary tree where each node knows the sum of the
+// weights of the children. This allows for addition, removal, and random
+// selection/dropping in logarithmic time.
+//
+// random(w) must be defined to return a uniform random value between zero
+// inclusive and w exclusive. The type W must support addition, binary and
+// unary -, and < comparisons, and W() must construct the zero value (these
+// constraints are met for primitive signed integer types).
+template <typename K, typename V, typename W, W random(W)>
+class WeightedMap
+{
+    struct Node {
+        K key;
+        V value;
+        W weight;
+        W sumOfDescendantWeights;
+    };
+
+    // The following vector is the tree representation of this collection.
+    // For each node, we keep track of the key, its associated value,
+    // its weight, and the sum of the weights of all its descendants.
+    std::vector<Node> nodes;
+
+    // The following map is to simplify removal.
+    std::map<K, size_t> indexMap;
+
+    static inline size_t leftChild(size_t i) { return i*2 + 1; }
+    static inline size_t rightChild(size_t i) { return i*2 + 2; }
+    static inline size_t parent(size_t i) { return (i-1)/2; }
+
+public:
+    // Check internal invariants (for tests).
+    void checkInvariants() const
+    {
+        assert(indexMap.size() == nodes.size());
+        for (size_t i = 0; i < nodes.size(); i++) {
+            assert(indexMap.at(nodes.at(i).key) == i);
+            assert(nodes.at(i).sumOfDescendantWeights == getWeightAt(leftChild(i)) + getWeightAt(rightChild(i)));
+        }
+    }
+
+private:
+    // Return the sum of weights of the node at a given index and all of its descendants.
+    W getWeightAt(size_t index) const
+    {
+        if (index >= nodes.size()) {
+            return W();
+        }
+        auto& node = nodes.at(index);
+        return node.weight + node.sumOfDescendantWeights;
+    }
+
+    // When adding and removing a node we need to update its parent and all of its
+    // ancestors to reflect its weight.
+    void backPropagate(size_t fromIndex, W weightDelta)
+    {
+        while (fromIndex > 0) {
+            fromIndex = parent(fromIndex);
+            nodes[fromIndex].sumOfDescendantWeights += weightDelta;
+        }
+    }
+
+    // For a given random weight, this method recursively finds the index of the
+    // correct entry. This is used by WeightedMap::takeRandom().
+    size_t findByWeight(size_t fromIndex, W weightToFind) const
+    {
+        W leftWeight = getWeightAt(leftChild(fromIndex));
+        W rightWeight = getWeightAt(fromIndex) - getWeightAt(rightChild(fromIndex));
```

Comment:
Why not move line 88 to line 93, so that you skip the unnecessary `getWeightAt` calls if `weightToFind < leftWeight`?

Reply:
Initially because I was trying to make the code close to the original. But yes, we've made other optimizations so we might as well make this one.